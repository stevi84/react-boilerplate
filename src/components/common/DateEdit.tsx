import { LocalizationProvider } from '@mui/x-date-pickers';
import { DatePicker } from '@mui/x-date-pickers/DatePicker';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
import { useField } from 'formik';
import { useTranslation } from 'react-i18next';
import { formatISO, isValid, parseISO } from 'date-fns';

interface DateEditProps {
  id: string;
  name: string;
  label: string;
}

export const DateEdit = (props: DateEditProps) => {
  const { id, name, label } = props;

  const [field, meta, helper] = useField(name);
  const { t } = useTranslation();
  const dateFormat = t('dateFormat');

  // konvertiert zwischen ISO8601-String (für Store) und Date (für Picker)
  const formatStore = (date: Date | null): string => (isValid(date) ? formatISO(date as Date) : '');
  const parseStore = (str: string): Date | null => {
    const date: Date = parseISO(str);
    return isValid(date) ? date : null;
  };
  const onChange = (newValue: Date | null) => helper.setValue(formatStore(newValue));

  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <DatePicker
        value={parseStore(field.value)}
        format={dateFormat}
        onChange={onChange}
        label={label}
        slotProps={{
          textField: {
            id,
            variant: 'outlined',
            error: !!meta.error,
            helperText: meta.error,
          },
        }}
        sx={{ margin: 1 }}
      />
    </LocalizationProvider>
  );
};
