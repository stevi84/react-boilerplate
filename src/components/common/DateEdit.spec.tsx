import { describe, it, expect, vi } from 'vitest';
import { createRenderer, ShallowRenderer } from 'react-test-renderer/shallow';
import { render, screen } from '@testing-library/react';
import { DateEdit } from './DateEdit';

const { setValueMock } = vi.hoisted(() => ({
  setValueMock: vi.fn(),
}));
vi.mock('formik', () => ({
  useField: () => [{ value: '2020-12-03T15:02:28+02:00' }, { error: 'error' }, { setValue: setValueMock }],
}));

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

describe('DateEdit', () => {
  it('should equal saved snapshot', () => {
    const renderer: ShallowRenderer = createRenderer();
    renderer.render(<DateEdit id="id" name="name" label="label" />);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should format the given value', () => {
    render(<DateEdit id="id" name="name" label="label" />);
    expect(screen.getByLabelText(/Dec 3, 2020/)).toBeTruthy();
    // expect(screen.getByLabelText('label')).toHaveValue('03at5For02at');
  });
});
