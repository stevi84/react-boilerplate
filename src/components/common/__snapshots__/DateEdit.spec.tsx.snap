// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateEdit > should equal saved snapshot 1`] = `
<LocalizationProvider
  dateAdapter={[Function]}
>
  <ForwardRef(DatePicker)
    format="dateFormat"
    label="label"
    onChange={[Function]}
    slotProps={
      {
        "textField": {
          "error": true,
          "helperText": "error",
          "id": "id",
          "variant": "outlined",
        },
      }
    }
    sx={
      {
        "margin": 1,
      }
    }
    value={2020-12-03T13:02:28.000Z}
  />
</LocalizationProvider>
`;
